src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,285 @@
+use std::sync::Arc;
+
+use common_daft_config::DaftExecutionConfig;
+use common_error::DaftResult;
+use daft_stats::TableMetadata;
+use parquet2::metadata::{FileMetaData, RowGroupList};
+
+use super::BoxScanTaskIter;
+use crate::{ChunkSpec, DataSource, ScanTask, ScanTaskRef};
+
+/// Performs splitting of an iterator of [`ScanTaskSplitByRowgroup`] into a flattened iterator of [`ScanTaskRef`].
+#[must_use]
+pub(crate) fn split_by_rowgroup<'a>(
+    splitters: impl Iterator<Item = DaftResult<SplittableScanTaskRef>> + 'a,
+    config: &'a DaftExecutionConfig,
+) -> BoxScanTaskIter<'a> {
+    Box::new(
+        splitters
+            .into_iter()
+            .flat_map(|scan_task_to_split| match scan_task_to_split {
+                Err(e) => EitherOnceOrMany::Once(std::iter::once(Err(e))),
+                Ok(scan_task_to_split) => {
+                    EitherOnceOrMany::Many(scan_task_to_split.into_iter(config).map(Ok))
+                }
+            }),
+    ) as BoxScanTaskIter
+}
+
+/// A ScanTask that can be split by Parquet rowgroups
+pub(crate) enum SplittableScanTaskRef {
+    NoSplit(ScanTaskRef),
+    Split(ScanTaskRef, FileMetaData),
+}
+
+/// Iterator for [`SplittableScanTaskRef`] which produces [`ScanTaskRef`]
+enum SplittableScanTaskRefIterator<'a> {
+    NoSplitIterOnce(ScanTaskRef, bool),
+    SplitIterMany(ParquetFileSplitter<'a>),
+}
+
+/// Iterator that performs a split on the provided [`ScanTaskRef`]
+struct ParquetFileSplitter<'a> {
+    // Iterator state that tracks iteration progress
+    current_idx: usize,
+    accumulated_rowgroup_ids: Vec<usize>,
+    estimated_accumulated_materialized_size: f64,
+
+    // Cached items
+    scan_task: ScanTaskRef,
+    file_metadata: FileMetaData,
+    config: &'a DaftExecutionConfig,
+
+    // Call once and cache this, since it can be expensive when called in a loop
+    scan_task_in_memory_size_estimate: Option<usize>,
+}
+
+impl SplittableScanTaskRef {
+    fn into_iter(self, cfg: &DaftExecutionConfig) -> SplittableScanTaskRefIterator {
```

Comment:
this should be under https://doc.rust-lang.org/std/iter/trait.IntoIterator.html

Then making an iterator should be automatic

Reply:
Propagated `&DaftExecutionConfig` through the structs and stored it on `self`, so I was able to convert this to an `IntoIterator` trait implementation